File: LeetCode/python3/347.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = collections.Counter(nums)
        min_heap = []
        for key, value in counter.items():
            if len(min_heap) < k or min_heap[0] < value:
                heapq.heappush(min_heap, value)
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        ret = []
        for key,value in counter.items():
            if value >= min_heap[0]:
                ret.append(key)
        return ret

File: LeetCode/cpp/347.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int> &nums, int k) {
        unordered_map<int, int> counter;
        for (auto &m: nums)
            ++counter[m];
        priority_queue<int, vector<int>, greater<>> heap;
        for (auto &freq:counter) {
            if (heap.size() < k || freq.second > heap.top())
                heap.push(freq.second);
            if (heap.size() > k)
                heap.pop();
        }
        vector<int> ret;
        for (auto &freq:counter) {
            if (freq.second >= heap.top())
                ret.push_back(freq.first);
        }
        return ret;
    }
};
